sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java

```diff
@@ -0,0 +1,118 @@
+
+    /**
+     * Creates StorageBearerTokenChallengeAuthorizationPolicy.
+     *
+     * @param credential the token credential to authenticate the request
+     * @param scopes the scopes used in credential, using default scopes when empty
+     */
+    public StorageBearerTokenChallengeAuthorizationPolicy(TokenCredential credential, String... scopes) {
+        super(credential, scopes);
+        this.scopes = scopes;
+    }
+
+    @Override
+    public Mono<Void> authorizeRequest(HttpPipelineCallContext context) {
+        return Mono.defer(() -> {
+            String[] scopes = this.scopes;
+            scopes = getScopes(context, scopes);
+            if (scopes == null) {
+                return Mono.empty();
+            } else {
+                return setAuthorizationHeader(context, new TokenRequestContext().addScopes(scopes));
+            }
+        });
+    }
+
+    @Override
+    public void authorizeRequestSync(HttpPipelineCallContext context) {
+        String[] scopes = this.scopes;
+        scopes = getScopes(context, scopes);
+
+        if (scopes != null) {
+            setAuthorizationHeaderSync(context, new TokenRequestContext().addScopes(scopes));
+        }
+    }
+
+    @Override
+    public Mono<Boolean> authorizeRequestOnChallenge(HttpPipelineCallContext context, HttpResponse response) {
+        return Mono.defer(() -> {
+            String authHeader = response.getHeaderValue(HttpHeaderName.WWW_AUTHENTICATE);
+            Map<String, String> challenges = parseChallenges(authHeader);
+
+            String scope = challenges.get("resource_id=");
+            if (scope != null) {
+                scope += DEFAULT_SCOPE;
+                scopes = new String[] { scope };
+                scopes = getScopes(context, scopes);
+                return setAuthorizationHeader(context, new TokenRequestContext().addScopes(scopes)).thenReturn(true);
+            }
+            return Mono.just(false);
+        });
+    }
+
+    @Override
+    public boolean authorizeRequestOnChallengeSync(HttpPipelineCallContext context, HttpResponse response) {
+        String authHeader = response.getHeaderValue(HttpHeaderName.WWW_AUTHENTICATE);
+        Map<String, String> challenges = parseChallenges(authHeader);
+
+        String scope = challenges.get("resource_id=");
+        if (scope != null) {
+            scope += DEFAULT_SCOPE;
+            scopes = new String[] { scope };
+            scopes = getScopes(context, scopes);
+            setAuthorizationHeaderSync(context, new TokenRequestContext().addScopes(scopes));
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Gets the scopes for the specific request.
+     *
+     * @param context The request.
+     * @param scopes Default scopes used by the policy.
+     * @return The scopes for the specific request.
+     */
+    public String[] getScopes(HttpPipelineCallContext context, String[] scopes) {
```

Comment:
Does this need to be public? I don't recall this being used anywhere outside of this class

Reply:
Taking this - had just based it off the other challenge policy in java https://github.com/Azure/azure-sdk-for-java/blob/6c4af651bda1ce4b3b81ebce67c2466fec9a0650/sdk/core/azure-core-management/src/main/java/com/azure/core/management/http/policy/ArmChallengeAuthenticationPolicy.java